pandas/core/dtypes/cast.py

```diff
@@ -462,6 +501,281 @@ def maybe_promote(dtype, fill_value=np.nan):
     return dtype, fill_value
 
 
+def maybe_promote_with_array(dtype, fill_value=np.nan):
```

Comment:
this is a huge additional to the technical debt. I am -1 on adding this at all. It is not at all clear whether this logic is correct and/or tested. more to the point, what is the purpose of all of this?

Reply:
@jreback you can pretty much ignore this PR; I'm asking @h-vetinari to keep it rebased for reference as we identify parts that are worthwhile to break off into bite-size pieces.

Reply:
> more to the point, what is the purpose of all of this?

There are a handful of places where we call maybe_promote where we _could_ have `fill_value` that is an ndarray.  Part of the plan for this is to identify in which of those cases we can rule out ndarray.

Reply:
i that’s fine

happy to pick off good changes

Reply:
> @jreback: this is a huge additional to the technical debt. I am -1 on adding this at all. It is not at all clear whether this logic is correct and/or tested. more to the point, what is the purpose of all of this?

Please read this comment, not just skim over it.

The array-path in `maybe_promote` is broken and both you and @jbrockmendel were excited to rip it out. At the same time, there's several potential or future use-cases for the array-case, and so I [asked](https://github.com/pandas-dev/pandas/pull/23982#issuecomment-541348431) [twice](https://github.com/pandas-dev/pandas/pull/25425#issuecomment-544366546) how this should be handled. 

Having a separate method is IMO the least invasive change, and would eventually still allow to rip out the array-path from `maybe_promote`. And more importantly, the logic is **tested with the same promotion tests**, which was the whole point of the `tests/dtypes/cast/test_promote.py`-module. Lastly, since it's a private method, there's no technical debt.

Reply:
>  Lastly, since it's a private method, there's no technical debt.

@h-vetinari i don’t even know what to say anymore

Reply:
Sorry for the misunderstanding, I meant as in API debt.

The technical debt is *already* there, in the array-path of `maybe_promote`. I'm trying to fix it. Feel free to address any of the comments or questions I've raised about this. But if you come into an ancient PR and - without regard for any of the existing context - assert that it must be garbage ("It is not at all clear whether this logic is correct and/or tested [it [is](https://github.com/pandas-dev/pandas/pull/25425#discussion_r340457853)]. more to the point, what's the purpose of all of this?"), then I'm gonna respond in kind.

Reply:
@h-vetinari pls don't respond like this. it is not helpful to anyone.

I can and will come into *every* PR and make comments. My purpose is to avoid cluttering pandas with technical debt. This PR just adds to it.